### ISSUE TITLE
.net library doesn't close down properly

### ISSUE THREAD

**Issue by trevorboultwood** [Grix/helios_dac#23]
I am using the .net helios library which works perfectly. However, after closing the program down the thread controlling the laser does not resulting in me having to close it within windows task manager.

**Comment by Grix** [comment]
There are no threads in the Helios .NET library as far as I can see... If so they are in the libusb library but that should all be freed when calling CloseDevices(). Are you sure this is a problem with the library and not a thread in your own program? If so, can you post a sample project demonstrating the issue so I can reproduce it?

**Comment by trevorboultwood** [comment]
Hi Grix, I have taken the following code from your example project. If I execute heliosController.OpenDevices(); when I close/press exit my process doesn't end. (have go into task manager) if i am debugging I have to press the stop button.

I am looking at implementing your driver into my existing .net project which is sadly a .net framework  project and requires porting to .net core. 

Thank you,

```
using HeliosDac;
using LibUsbDotNet.Main;
using LibUsbDotNet;
using LibUsbDotNet.WinUsb;

namespace WinFormsApp1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        UsbRegistry usbRegistry;
        HeliosController heliosController = new HeliosController();
        private void Form1_Load(object sender, EventArgs e)
        {
            int numberOfDevices = 0;
            try
            {
                numberOfDevices = heliosController.OpenDevices();
                MessageBox.Show("Number of devices: " + numberOfDevices);
                for (int deviceId = 0; deviceId < numberOfDevices; deviceId++)
                {
                    MessageBox.Show(heliosController.GetName(deviceId));
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show("Failure during detecting and opening of Helios DACs: " + ex.Message);
            }
            heliosController.CloseDevices();
        }
    }
}
```

**Comment by trevorboultwood** [comment]
I should also mention, currently I do not have my Helios product connected. Will be purchasing soon just in the middle of a house move so cant confirm delivery address.

**Comment by Grix** [comment]
What code does it say is running when you pause / "break all" in the debugger?

**Comment by trevorboultwood** [comment]
![capture](https://user-images.githubusercontent.com/22022831/227798502-aaec8f9b-9673-4a63-9725-939cdf00f65c.png)

Hi,

.NET ThreadPool Gate & .NET system events. Apologies, this is not my main area of expertise. Thank you for supporting me on this issue.

**Comment by Grix** [comment]
I am unable to reproduce the problem. I create a new winforms app, copy in your code, run it, then close the window. Everything closes down normally, nothing stays running. Do you have anything else in your app?

**Comment by trevorboultwood** [comment]
> I am unable to reproduce the problem. I create a new winforms app, copy in your code, run it, then close the window. Everything closes down normally, nothing stays running. Do you have anything else in your app?

Hi Grix, thank you for testing for me. I have have just removed all my installations of visual studio to try and get this to work and have tried a variety of project settings. Would you be able to send me over your project so i can see any differences? Apologies for the inconvenience.

I don't have anything else running my app, I created a WindowsFormApp and include the HeliosDac assembly and place the lib dlls in the windows-bin folder.

**Comment by trevorboultwood** [comment]
As well as the HeliosDac.dll I also add the following com references: LibUsbDotNet & LibUsbDotNet.LibUsnDotNet

**Comment by Grix** [comment]
Here is my project that works ok.
[WinFormsApp1.zip](https://github.com/Grix/helios_dac/files/11087240/WinFormsApp1.zip)

**Comment by trevorboultwood** [comment]
HI Grix, thank you for sharing the project. I think however you have made a slight error. The function handler which you have copied never gets executed.

I have double click the windows form from within the designer view and it generates a new handler where I then placed the original code in.

Sadly the issue persists. I took another look last night and didn't have any luck. Thank you again.

 ```
private void Form1_Load(object sender, EventArgs e)
        {

        }

        private void Form1_Load_1(object sender, EventArgs e)
        {

            int numberOfDevices = 0;
            try
            {
                numberOfDevices = heliosController.OpenDevices();
                MessageBox.Show("Number of devices: " + numberOfDevices);
                for (int deviceId = 0; deviceId < numberOfDevices; deviceId++)
                {
                    MessageBox.Show(heliosController.GetName(deviceId));
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show("Failure during detecting and opening of Helios DACs: " + ex.Message);
            }
            heliosController.CloseDevices();


        }
```

**Comment by Grix** [comment]
Can you copy your entire project and send it to me then.

**Comment by trevorboultwood** [comment]
[tboultwood_WFA_Helios.zip](https://github.com/Grix/helios_dac/files/11092985/tboultwood_WFA_Helios.zip)

The only notable difference between the WFA you supplied and the one I have been using on my side is you installed the NuPackage and I directly referenced the .dll supplied on your repo for LibUsbDotNet.

**Comment by trevorboultwood** [comment]
Hi Grix, I have continued with some debugging and I have discovered the bug.

`UsbDevice.ForceLibUsbWinBack = true;`

This causes the application to hang and commenting it out allows the application to function correctly.

Whilst I have no hardware to test with just yet I found calling this function solves the issue. I assume this should be included in the closedevices function.

`UsbDevice.Exit();`

Thank you for providing me with assistance. I might purchase the Helios DAC to a relatives house who can forward it on to me.
